### ISSUE TITLE
Hydra v2 Progress Issue

### ISSUE THREAD

**Issue by bedeho** [Joystream/hydra#10]
# `Mon, Aug 24th`

## Agenda

- What are our priorities now that Metin is back and Hydra has been submitted?
- I had suggested a set of focus areas that needed to be reviewed, and work had to be split up
```
a) faster processing/synching
b) decoupling blockchain synching and processing, so that one can easily rerun processing when altering a schema or mapping.
c) static type safety in all mappings <== this latter point needs feasibility input from Mokhtar, but I am quite sure its possible. If its possible, I think the upside is substantial enough.
d) more integration tests on Hydra

I think Arsen can start as soon as he is ready, even if these are not done, but let me know if you think that would be counter-productive.
```

## Present

- Metin
- Dmitrii
- Bedeho

## Topics covered
- What is Dmitrii currently working on?
- What Metin is working on, and details of how to address those bugs.
- Do we really need to fix the mappings for Kusama treasury now that we have submitted, its not the highest priority?
- Should we continue to try to fix the out of memory issue now?
- Perhaps we need a better solution for handling naming conflicts, using a manifest or some other more explicit approach.

## Conclusions
- Dmitrii will focus on a+b, and mix in d for the next week or so.
- Metin will focus on writing mappings, with tests for some part of our runtime, and will try to identify bugs and rough edges of the developer workflow. Its very important here to get to a place where we find out how to give mapping author confidence that they are doing things correctly.
- We will delay and see what to do about c, hopefully we can settle next meeting.
- We will delay any work on manifest solution for now, Dmitrii will make issue.
- The out of memory bug will either implicitly get resolved by Dmitriis work, or it will pop up again in our own node, and then we will have better shot at local reproduction.

**Comment by bedeho** [comment]
# `Mon, Sep 7th`

## Agenda

1. Review [Babylon network release plan](https://github.com/Joystream/joystream/issues/1249), and resolve key questions
  - How to do testing and development workflow
  - What scope of work should be included in the release
  - What is the current status
2. How to split the work going forward on the release.
3. Is there a way to deal with [pagination issue](https://github.com/Joystream/joystream/issues/1146#issuecomment-686397471) identified by Atlas team without waiting for Warthog.

## Present

- Metin
- Dmitrii
- Bedeho

## Conclusions

1. We will need input from Mokhtar on how to do effective integration testing of the query node. We were concerned that running it as part of the overall testing framework would be a hassle, for example in terms of running time. Having standalone integration tets just for the query node would allow us to have a prebuilt node and only run certain queries. But then it does feel like a lot to have both network and node integration tests.

2. We decided against including any of the new advanced work, such as typed mappings or manifest files. First off, there is just the uncertainty around feasibility and timeline. Secondly, as emphasized by Dmitrii, we may need to refactor the entire Hydra framework to be more robust, with an architecture similar to Subscan/Polkascan. We will revisit this later.

3. Metin will start working on the schemas and mappings, as Arsen has been delayed, and Dmitrii will try to work on Hydra itself.

4. There was a polkadotjs version incompatibility issue which made it hard to smoothly populate the indexer database. The only easy way to resolve is to have access to a newer versions of Polkdotjs from what Dmitrii and Metin can gather. Dmitrii will try to experiment with a new version and synching against the new Joystream node for the imminent Alexandria release, and we will seek advice from Mokhtar as well.

**Comment by bedeho** [comment]
# `Tue, Sep 7th`

## Agenda

1. What is required to make the query node part of network testing infrastructure?

2. How can we turn off running irrelevant scenarios when working on PRs that are focused only on query node?

3. How can we avoid rebuilding full node & runtime when working on PRs that are focused only on query node?

4. What is the current status of the indexer issued with Polkadotjs

## Present

- Metin
- Dmitrii
- Mokhtar
- Bedeho

## Conclusions

1. @mnaamani  can solve all query node related requirements so long as we have query node in docker image, which we do. Ansible can be used to run with correct prebuilt assets and limit scenarios by looking at either labels or commit message. This can all be handled by Mokhtar

2. @dzhelezov found that with polakdotjs 1.31 he was able to run indexer against Kusama and at that quickly. He believes this now works is because of this new feature https://github.com/polkadot-js/api/pull/2535. However, we need to confirm whether this version of polkadotjs is compatible with our current Substrate version. He will ask around to figure out that compatibility. If that is compatible, we can just keep using this version for Hydra/query node exclusively, no other part of the Joystream code base needs to upgrade. If this is not compatible, then we need to consider alternatives for the indexer that are not based on polkadotjs high level APIs.

**Comment by bedeho** [comment]
# `Mon, Sep 14th`

## Agenda

1. What is our status on the release so far?

2. We have had some contradictory mental models, at least it appeared, on how exactly the architecture of the processor & indexer should be, summarized in this question issue from Bedeho: https://github.com/Joystream/hydra/issues/24#issuecomment-691445871

## Present

- Metin (@metmirr) 
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)

## Covered

1. The difference in the model was that Dmitrii was foreseeing lots of distinct processors within the Joystream API, while I was thinking there was just one. Distinct processors, generating distinct APIs, which then can be stitched together at a top-level API, is sort of in GraphQL spirit, however, it wasn't clear that the underlying Joystream Runtime data model. and final AP actually could work with this sort of segmentation, and it appeared to cause many complexities. We do however need to look out for ways of making the schemas manageable to work with, even if there is one monolithic query state.

2. The estimation went reasonably well for both Dmitrii & Metin, however, there was some significant uncertainty about some parts of the work for both.

2. Can we run the indexers sooner rather than running everything later? There are still lurking bugs and issues showing up with the Kusama network, and we are better off trying to get a node running as soon as possible.

3. How will Arsen (@iorveth)  & Mokhtar (@mnaamani ) contribute to various Hydra & query node tasks in Babylon?

## Conclusions

1. We will only have a single monolithic processor and query stage for the query node, this impacts some of Dmitrii's tasks, which he will update.

2. We will attempt to get an indexer running as soon as possible, and we will try to have a long-running indexer talking to a simple Babylon full node as soon as we can, just producing semi-empty blocks. 

3. Mokhtar will assist in updating Polakdotjs joystream types library as soon as possible, as Lezsek will not be back for at least a week. It's not clear how much of the query-node integration testing scenarios he will write, but he may contribute once he is done with his main task of integrating the query node in the testing infrastructure itself. To begin with, all tasks are assigned to Dmitrii.

4. Dmitrii will add a separate task about writing integration tests just for the Hydra indexer.

5. Arsen will work with Metin to write the query-node when he is done with a working runtime, as he is very knowledgeable about the content directory, but initially, all tasks are assigned to Metin.

**Comment by bedeho** [comment]
# `Mon, Sep 21th`

## Agenda

1. Where are we, where can we be next week?
2. Discuss: https://github.com/Joystream/joystream/issues/1409
        
## Present

- Arsen (@iorveth)
- Metin (@metmirr) 
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)

## Covered

1. Agenda
  - Arsen has gotten started and reviwed his first query node PRs.
  - Metin is close to finishing the input schema, and will share in next few days.
  - Dmitrii has been working on integrating processor and indexer
2. How to proceed the next week.
3. Performance issues with indexer API due to default GraphQL resolver behaviour in Warthog (which also powers this API) to not consolidate queries.
4. Substrate builders program application timing: we discussed when to apply, and also what to do about funding further Hydra development in general
5. Hydra future.

## Conclusions

1. Arsen and Metin will focus on writing mappings the next week, but it's unclear how much progress can be made.

2. Dmitrii will focus on completing the integraiton of the indexer and processor, and then commence with first Hydra integration test on indexer with template chain.

3. We will postpone a proper fix of the API issue until after Babylon is out.

4. We should do the builders program ASAP, Dmitrii will look into what is required to apply, and the timeline and commitments required of the program. We should also apply to other funding sources, but this requires harmonizing requests guided by an underlying plan and vision for what we want to do. This planning could take a little more time. Bedeho is happy to spend time on this, but needs some calendar space to get started.

5. This is a complex topic, we should schedule explicit time for it later. It does not appear to be a rush, as any plan depends on what we are doing anyway, which is make Hydra better for anyone to use.

**Comment by dzhelezov** [comment]
## Quick recap of Hydra as of 29.09 ##

- The indexer (ingests blocks and stores in the database) publishes updates in real-time to Redis
- The Indexer API (a GraphQL endpoint powered by Warthog) exposes events and extrinsics for querying 
- The processor can ingest the events of interest from the Indexer API via polling

TBD Core Hydra features
Priority: 1
- Replace the processor polling with websocket subscriptions to events of interest.  It is broken down as follows:
   - The Indexer API exposes a GraphQL subscription <- currently blocked by https://github.com/goldcaddy77/warthog/issues/422
   - Update the processor ingestion logic
- Improve the performance of the API requests: not necessary, but nice to have
- (Optional) Put more metrics into Redis

TBD Pre-integration tests (w/o mappings logic)
Priority: 2
- Mock events & extrinsics data
- Test indexer ingestion
- Test processor subscriptions

TBD Infrastructure tasks:
Priority: 3
- Basic Prometheus exporter

**Comment by bedeho** [comment]
# `Wed, Sep 30th`

## Agenda

1. Review https://github.com/Joystream/hydra/issues/10#issuecomment-700580784 
2. Dmitrii status on: integration tests & subscriptions
3. Metin status on: schema progress
4. Dmitrii: re-evaluation of monitoring & cost of introducing type safe mappings
5. Can we get transaction handlers?
        
## Present

- Metin (@metmirr) 
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)

## Covered

1. Dmitrii explained the role of the Redis message broker, which is as a queue between the indexer node and the indexer API server. The API will provide a GraphQL subscription which alerts client when a new block has been fetched, and the client would at this point fetch the block using a separate query.

2. There is a limitation in Warthog preventing the completion of the approach described in the prior point, but it is expected that this will be addressed by the Warthog maintainer with a small fix soon, however, if that was to fall through we can rely on the already working polling based approach in the processor.

3. Metin is working on mappings, of which we expect 25 or so to be needed, and it takes about 2-3 hours per small group  (2-3) of related events. Arsen has not started contributing on the mapper side yet.

4. We discussed how type safe mappings would be implemented, and how to deal with the fact that the Polkadot/Joystream types are distinct from the Warthog data model types, and also how to deal with the fact that recovering extrinsic parameters when processing events can be quite hard, and its not clear how to make type safe mappings in such cases. The conclusion here was that we need transaction handlers to sidestep this entirely.

5. We briefly discussed when a working chain for Babylon with relevant transactions.

6. How should query deployment and hosting be handled?

## Conclusions

- Metin will write mappers segmented into a pre-handler, which takes `SubstrateEvents`, and a core handler, which takes static types. This distinction will allow us to make a clean transition later when we only have core handlers.

- Dmitrii will update the project board to reflect a focus on the following: proceed to focus on Hydra integration tests using a Substrate template chain with very plain transactions, such as transfers, validation, etc. Whenever the Warthog issue is resolved, he will shift focus and complete that work.

- We cannot introduce transaction handlers for Babylon, its too risky.

**Comment by bedeho** [comment]
# `Mon, Oct 5th`

## Agenda

- Dmitrii 
  - status?
  - how is integration testing going?
  - Warthog blocker?
  - Weeks goal: 
- Metin
  - status?
  - how are high level schemas going
  - how is prehandler approach working out?
  - ID field issue
  - where can Arsen start?
  - Weeks goal?
- Arsen
  - Weeks goal: 
- Get synched on Major directions:
https://github.com/Joystream/hydra/issues/8
        
## Present

- Metin (@metmirr)
- Arsen (@iorveth)
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)

## Covered

1. Dmitrii has sidestepped performance issue which demanded use of subscriptions to keep up with index status, hence we are no longer depending on Warthog fix in time, but we will make switch whenever fix is available.
2. Dmitrii has made initial lower level integration tests to ensure that indexer worked, and they are passing.
3. Dmitrii suggested we introduce deeper e2e logging infrastruture for all of our hosted infrastructure, and that this perhaps would make sense to bundle with Kubernets introduction.
4. Metin has reworked input schemas to only have higher level types.
5. Metin has written prehandlers for membership queries, and so far it looks clean.
6. We disussed how to deal with ID fields in input schemes, and what changes to Warthog schema generation code would be required.
7. We were all synched on Major directions issue.

## Conclusions

- By end of week Dmitrii will try to get e2e test working with an actual substrate chain.
- By end of week Metin & Arsen will together attempt to cover all mappings.
- Metin will change ID handling in schema generation.

**Comment by bedeho** [comment]
# `Mon, Oct 12th`

## Agenda

This was an impromptu meeting to discuss a single urgent issue of mismatching expectations between testing and hydra/query node teams. There was no sufficient time to properly do a team call today.
        
## Present

- Metin (@metmirr)
- Arsen (@iorveth)
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)
- Mokhtar (@mnaamani)

## Covered

1. Mokhtar explained that, he was working on refactoring integration testing code, and there was still multiple steps remaining in order to just build and run the query node as part of the CI. He was not sure whether it was still on his task list to write the actual integation tests, and he needed help from Dmitrii to get query node up and running.
2. Dmitrii was mainly focusing on getting e2e tests with a chain to run, something which he was nearing in on.
3. Metin & Arsen were still focusing on mappings, and in particular tackling what should be used for Ids in various entities.
4. We also discussed how we could shift resource around to assist Mokhtar

## Conclusions

- Dmitrii will keep going on his own, and help Mokhtar getting the query node to work.
- Arsen will leave working on mappings at the earliest possible time, and take charge of writing tests. Possibly with Mokhtar, depending on when he will be able to get to that.
- There will be a second call this week where Metin+Arsen+Bedeho can resolve the ID issue, and also how to get Arsen to transition.

**Comment by bedeho** [comment]
# `Tue, Oct 13th`

## Agenda

This was an impromptu meeting to discuss a single urgent issue of mismatching expectations between testing and hydra/query node teams. There was no sufficient time to properly do a team call today.
        
## Present

- Metin (@metmirr)
- Arsen (@iorveth)
- Bedeho (@bedeho)
- Leszek (@Lezek123)

## Covered

1. The nature of the prior entity ID problem: 
  - Warthog had conflicting id with input schema id, so the latter had to be ignored, this has been fixed.
  - Allowing mapping author to provide id field value required hydra CLI changes, this has been done.
2. We discussed Babylon test plan proposal: https://github.com/Joystream/joystream/issues/1526
  - We can already create classes and schemas based on JSON using tooling.
  - OK, Arsen will provide feedback, or request assistance if there is any problem.
3. How to handle property value update events
  - Only easy approach seems like static mapping from IDs to Warthog data model fields
  - Leszek may offer extra information about how he does the opposite

## Conclusions

- Arsen 
  - Create new tasks for making integration tests, look at Mokhtar issues.
  - Get familiar with new docs that Leszek is wiriting.
  - Will try to finish trait related problem  in runtime.
- Metin
  - Property update mappings
  - Add support for pagination
  - When Metin is done with tasks, he may be able to help with testing effort.
- Leszek
  - Add more docs/examples for how to use your tooling as a library for making content directory transactions, really specifically about: adding entity, adding schema support to entity, updating property
  - Will create input JSON files, and introduce tooling in CI.

**Comment by bedeho** [comment]
# `Tue, Oct 13th`

## Agenda

This was an impromptu meeting to discuss a single urgent issue of mismatching expectations between testing and hydra/query node teams. There was no sufficient time to properly do a team call today.
        
## Present

- Metin (@metmirr)
- Arsen (@iorveth)
- Bedeho (@bedeho)
- Leszek (@Lezek123)

## Covered

- Current status for Arsen: he is investigating writing test scenarios.
- Current status of Metin he is completing mappings
- Current status of Dmitrii: he is working on hosted query node for staging network
- Goals for next week for everyone

## Conclusions

- Dmitrii will focus on
  - improving documentation that helps @mnaamani , based on his feedback
  - introducing Hydra integration tests in Hydra repo CI
  - starting to cleanup tech debt from babylon, which includes separating libraries and workspaces for maintainability. This is a prerequisite for any new features and maintainability in the future.
  - refactoring Hydra CLI to be more useful to developers
- Arsen + Metin have the combined goal of completing most testing scenarios this week.

**Comment by bedeho** [comment]
# `Tue, 27th Oct`

## Agenda

This was a weekly meeting to discuss progress on Hydra
        
## Present

- Metin (@metmirr)
- Arsen (@iorveth)
- Bedeho (@bedeho)
- Leszek (@Lezek123)

## Covered

- Current status for Arsen: he is investigating writing test scenarios, estimates to be about **30%** done, but is still waiting for @mnaamani to allow Polkadotjs to be usable
- Current status of Metin: he is completing mappings for the `transaction` type.
- Current status of Dmitrii: he is extending hydra indexer status information to allow richer status & failure monitoring
- An outsanding issue about how BigIntegers are handled across the query API and database. It was not entirely clear what the state of affairs was currently, but Metin will look into it.
- Goals for next week for everyone

## Conclusions

- Dmitrii will focus on
  - completing introduction of new status endpoints in indexer
  - writeup an issue describing the set of refactorings he will tackle next.
  - start on those refactorings.
- Metin will complete mappings, which should take 3-6h, and then focus on tackling how BigIntegers are being tackled.
- Arsen will continue writing integration tests.

**Comment by bedeho** [comment]
# `Mon, 2nd Nov`

## Agenda

This was a weekly meeting to discuss progress on Hydra.
        
## Present

- Dmitrii (@dzhelezov)
- Metin (@metmirr)
- Arsen (@iorveth)
- Bedeho (@bedeho)

## Covered

- Current status for Arsen: He will be submitting tests today, hopefully, covering phase 1 of tests. There was however a blocking error. We also identified the issue that these tests currently do not hook into query node progress before querying for sideffect of doing transactions with the chain, hence this must be fixed. Dmitrii confirmed that the currently used Hydra-Cli alrady supports this. This progress API is separate from the normal API, and the dockr-compose file will show which port one should connect to on the same host. He has not yet started integrating anything with storage system, @Lezek123 will be a great resource on this.
- Current status of Metin: Been occupied with some reviews in transition between mappings and starting on tests.
- Current status of Dmitrii: Working on week long refactoring plan, described here https://github.com/Joystream/hydra/issues/82

## Conclusions

- Dmitrii will continue focusing on refactoring.
- Metin will join Arsen to work on tests.
- Arsen will continue writing integration tests, hopefully both of them will complete phase 2 together this week.

**Comment by bedeho** [comment]
# `Mon, 2nd Nov`

## Agenda

This was a weekly meeting to discuss progress on Hydra.
        
## Present

- Mokhtar (@mnaamani)
- Dmitrii (@dzhelezov)
- Arsen (@iorveth)
- Bedeho (@bedeho)

## Covered

- Mokhtar was briefly part of the meeting, asking about how we do migrations and database setup for Hydra.
- Current status for Arsen: Is about to wrap up work on phase 3 integration tests, will leave storage system integration to Metin.
- Dmitrii status: Is basically done with main refactoring, is not working on getting integration tests working again, and then will move on to triaging new feature improvements. We discussed what to focus on next, Bedeho expressed a priority for type safe mappings, both for transactions and events, and then support for meta abstractions like smart contract modules. Dmitrii was interested in understanding how we can improve things for deployment, which currently has big inadequacies, and also learning more from Metin & Arsen about what pain points they felt.

## Conclusions

- Dmitrii will fix tests, review feedback from Mokhtar+Arsen+Metin,and then we will have a meeting about what to prioritise.
- Arsen will wrap up tests and move on to other things for now, and write about obstacles.

**Comment by bedeho** [comment]
# `Mon, 16th Nov`

## Agenda

This was a weekly meeting to discuss progress on Hydra.
        
## Present

- Metin (@metmirr)
- Dmitrii (@dzhelezov)
- Bedeho (@bedeho)

## Covered

- What to do about various discrepancies identified by the Atlas team API expectations and actual API of query node described here: https://github.com/Joystream/joystream/issues/1698
- Review Hydra backlog issue for future work on Hydra prepared by Dmitrii: https://github.com/Joystream/hydra/issues/109

## Conclusions

- We will fix the problem of confusing auto-generated Warthog fields bieng introduced in the API in the future: https://github.com/Joystream/hydra/issues/117
- Our main priorities for new features in Hydra are on the processor side, and Dmitrii will attempt to come up with a design doc for how all of the different features envisioned will work together.
- @Lezek123 will change `channel.title` to `channel.handle` in the schemas and CLI.
- Metin will
  - look into `MediaLocation` typing.
  - talk to Klauiudz about nullbiloity of cover photo URLs.
  - look into non-nullability of relationship types in order to fix.
  - look into discrepancy between property type in storage and event for transactions.
  -  fixes name of full text search.
- Arsen will fix integration testing based on what Metin changes.
- Dmitrii will
   - fix name of autogenerated types using better plural forms.
   - add issues for introducing e2e test cases that would cover the bugs recently identified by Klauiusz.
   - will write a design doc for how all the different changes on the processor side fit together.

**Comment by bedeho** [comment]
Closing this now, as we are basically done with everyhting for Babylon and v2.
